tls/s2n_early_data_io.c

```diff
@@ -117,3 +119,135 @@ S2N_RESULT s2n_early_data_validate_recv(struct s2n_connection *conn)
     RESULT_ENSURE(s2n_conn_get_current_message_type(conn) == END_OF_EARLY_DATA, S2N_ERR_EARLY_DATA_NOT_ALLOWED);
     return S2N_RESULT_OK;
 }
+
+static bool s2n_early_data_can_continue(struct s2n_connection *conn)
+{
+    uint32_t remaining_early_data_size = 0;
+    return s2n_connection_get_remaining_early_data_size(conn, &remaining_early_data_size) >= S2N_SUCCESS
+            && remaining_early_data_size > 0;
+}
+
+S2N_RESULT s2n_send_early_data_impl(struct s2n_connection *conn, const uint8_t *data, ssize_t data_len,
+        ssize_t *data_sent, s2n_blocked_status *blocked)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(blocked);
+    *blocked = S2N_NOT_BLOCKED;
+    RESULT_ENSURE_REF(data_sent);
+    *data_sent = 0;
+
+    RESULT_ENSURE(conn->mode == S2N_CLIENT, S2N_ERR_SERVER_MODE);
+
+    if (!s2n_early_data_can_continue(conn)) {
+        return S2N_RESULT_OK;
+    }
+
+    /* Attempt to make progress in the handshake even if s2n_send eventually fails.
+     * We only care about the result of this call if it would prevent us from calling s2n_send. */
+    int negotiate_result = s2n_negotiate(conn, blocked);
+    if (negotiate_result < S2N_SUCCESS) {
+        if (s2n_error_get_type(s2n_errno) != S2N_ERR_T_BLOCKED) {
+            return S2N_RESULT_ERROR;
+        } else if (*blocked != S2N_BLOCKED_ON_EARLY_DATA && *blocked != S2N_BLOCKED_ON_READ) {
+            return S2N_RESULT_ERROR;
+        }
+    }
+    /* Save the error status for later */
+    int negotiate_error = s2n_errno;
+    s2n_blocked_status negotiate_blocked = *blocked;
+
+    /* Attempt to send the early data.
+     * We only care about the result of this call if it fails. */
+    uint32_t early_data_to_send = 0;
+    RESULT_GUARD_POSIX(s2n_connection_get_remaining_early_data_size(conn, &early_data_to_send));
+    early_data_to_send = MIN(data_len, early_data_to_send);
+    if (early_data_to_send) {
+        ssize_t send_result = s2n_send(conn, data, early_data_to_send, blocked);
```

Comment:
Is it possible that this might send just the tail of the early data (as normal application data after the handshake completed successfully above) after the server rejected some previously sent early data?

Reply:
No. `early_data_to_send` will be 0 if the server rejected early data, so we won't send. https://github.com/aws/s2n-tls/blob/61b996625635847048449ffc6b2bf3f2969cd181/tests/unit/s2n_early_data_io_api_test.c#L477-L517